File: src/1481-Least-Number-of-Unique-Integers-after-K-Removals/least-number-of-unique-integers-after-k-removals.py
from typing import List, Dict

class Solution:
   def findLeastNumOfUniqueInts(self, arr: List[int], k: int) -> int:
        frequency_map: Dict[int, int] = {}

        for num in arr:
            if num in frequency_map:
                frequency_map[num] += 1
            else:
                frequency_map[num] = 1

        sorted_frequency: Dict[int, int] = sorted(frequency_map.items(), key=lambda x: x[1])
        
        while k > 0:
            key, count = sorted_frequency[0]
            if count > 1:
                sorted_frequency[0] = (key, count - 1)
            else:
                del sorted_frequency[0]
            k -= 1

        
        return len(sorted_frequency)


if __name__ == "__main__":
    arr: List[int] = [4,3,1,1,3,3,2]
    k: int = 3
    solution = Solution()
    
    print(solution.findLeastNumOfUniqueInts(arr, k))

File: src/1481-Least-Number-of-Unique-Integers-after-K-Removals/least-number-of-unique-integers-after-k-removals.js
/**
 * @param {number[]} arr
 * @param {number} k
 * @return {number}
 */
var findLeastNumOfUniqueInts = function(arr, k) {
  const frequency_map = new Map();

  for (const num of arr) {
    frequency_map.set(num, (frequency_map.get(num) || 0) + 1)
  }

  const sorted_frequency = [...frequency_map.entries()].sort((a, b) => a[1] - b[1])

  while (k > 0) {
    const [key, count] = sorted_frequency[0]
    if (count > 1) {
      sorted_frequency[0] = [key, count - 1]
    } 
    else {
      sorted_frequency.shift()
    }
    k -= 1
  }

  return sorted_frequency.length
};

arr = [4,3,1,1,3,3,2]
k = 3

console.log(findLeastNumOfUniqueInts(arr, k))